### DISCUSSION TITLE
OAuth2 for Tools

### DISCUSSION THREAD

**Discussion by wayum999** [danny-avila/LibreChat#3773] [Feature Requests & Suggestions]
I might be missing something and I am sorry if it is repeated, but why is the OAuth option in the OpenAPI tool setup greyed out? Is that available and I am missing something? Or is that being restricted to custom tools/plugins only? Thats probably one of the most useful features for me in OpenAI's GPT version because it handles the authentication  beautifully when OAuth is needed. I see many OAuth settings for login but not this one. Thanks for any help you can give, love the app!

**Comment by danny-avila** [comment]
Hey @wayum999 I've just not gotten a chance to set this up and test it properly. 

If you'd like to help me with an openapi spec, I'd be glad to get it implemented.

**Reply by wayum999** [reply]
really thats awesome! so here is a very basic openapi spec for github. the only function on it is to get the authenticated user but it will show you if its working or not. you will also need to setup a github oauth app and set the callback URL properly, if you dont know how to do you just go to your github settings > developer settings > oauth apps and create one there to get the client ID and secret and set the callback URL. 

This is a big one for me with the assistants (and hopefully any other tool LLM) because so many possibilities exist with all kinds of interactions and context when you can handle OAuth. Looking forward to it, let me know if you need anything else.

Also one other request I had was integration with Coze as it is a truly superior builder despite its appearance at first glance. I actually started making an adapter via python but now am using express due to issues i had with streaming a streaming response, also getting held up by the differences between how it works vs openai, namely the multi-modal handling. Similar, but different. Anyways, I am going to keep working on it and will let you know if i complete so people can use if they want. It was a bit much for me to dive into making a new LLM in your core code and I didn't want to mess too much with it and have a negative effect on updating so was compartmentalizing it separately.  Perhaps if I do get it to work someone or you could implement into the core. 

Anyways, thanks much.

openapi: 3.1.0
info:
  title: GitHub API - Get Authorized User
  version: 1.0.0
  description: This API specification includes OAuth 2.0 authentication to retrieve the authorized user's information using the GitHub API.

components:
  securitySchemes:
    OAuth2:
      type: oauth2
      flows:
        authorizationCode:
          authorizationUrl: https://github.com/login/oauth/authorize
          tokenUrl: https://github.com/login/oauth/access_token
          scopes:
            read:user: Grants read access to a user's profile data.
  schemas:
    User:
      type: object
      properties:
        login:
          type: string
          description: The user's login name.
        id:
          type: integer
          description: The user's unique ID.
        avatar_url:
          type: string
          description: The URL of the user's avatar.
        name:
          type: string
          description: The user's full name.
        email:
          type: string
          description: The user's email address.
        created_at:
          type: string
          format: date-time
          description: The date and time when the user's account was created.

security:
  - OAuth2:
      - read:user

paths:
  /user:
    get:
      summary: Get the authenticated user's profile
      description: Retrieves the profile information of the authenticated user.
      operationId: getAuthenticatedUser
      security:
        - OAuth2:
            - read:user
      responses:
        '200':
          description: A successful response containing the user's profile information.
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/User'
        '401':
          description: Unauthorized, invalid or expired OAuth token.
        '403':
          description: Forbidden, insufficient scope.

servers:
  - url: https://api.github.com
    description: GitHub API

**Comment by wayum999** [comment]
hey there, i know you have been busy, but was just wondering if there was an oauth option on the horizon for the agents or openai assistants for their tools? i see its still greyed out. thanks for everything, the app is amazing!

**Reply by rubentalstra** [reply]
hi, I have started with implementing OAuth for actions how this is also done by openAI: #5655

**Reply by rubentalstra** [reply]
![Screenshot 2025-02-05 at 11 36 57](https://github.com/user-attachments/assets/a8d009bd-fb24-4b7d-9c4f-b54c744e6ca5)
